openedx/features/lti_course_tab/tab.py

```diff
@@ -0,0 +1,247 @@
+"""
+Code related to LTI course tab functionality.
+"""
+from typing import Dict
+from urllib.parse import quote
+
+from django.contrib.auth.models import AbstractBaseUser
+from django.contrib.sites.shortcuts import get_current_site
+from django.http import HttpRequest
+from django.utils.translation import get_language, to_locale, ugettext_lazy
+from lti_consumer.lti_1p1.contrib.django import lti_embed
+from lti_consumer.models import LtiConfiguration
+from opaque_keys.edx.keys import CourseKey
+from web_fragments.fragment import Fragment
+
+from lms.djangoapps.courseware.access import get_user_role
+from lms.djangoapps.courseware.tabs import EnrolledTab
+from openedx.core.djangoapps.discussions.models import DiscussionsConfiguration
+from openedx.core.djangolib.markup import HTML
+from common.djangoapps.student.models import anonymous_id_for_user
+from xmodule.course_module import CourseDescriptor
+from xmodule.tabs import TabFragmentViewMixin, key_checker
+
+
+class LtiCourseLaunchMixin:
+    """
+    Mixin that encapsulates all LTI-related functionality from the View
+    """
+
+    ROLE_MAP = {
+        'student': 'Student',
+        'staff': 'Administrator',
+        'instructor': 'Instructor',
+    }
+    DEFAULT_ROLE = 'Student'
+
+    def _get_additional_lti_parameters(self, course: CourseDescriptor, request: HttpRequest) -> Dict[str, str]:
+        lti_config = self._get_lti_config(course)
+        additional_config = lti_config.lti_config.get('additional_parameters', {})
+        return additional_config
+
+    @staticmethod
+    def _get_user_id(user: AbstractBaseUser, course_key: CourseKey):
+        return anonymous_id_for_user(user, course_key)
+
+    def _get_lti_roles(self, user: AbstractBaseUser, course_key: CourseKey) -> str:
+        return self.ROLE_MAP.get(
+            get_user_role(user, course_key),
+            self.DEFAULT_ROLE,
+        )
+
+    @staticmethod
+    def _get_context_id(course_key: CourseKey) -> str:
+        return quote(str(course_key))
+
+    @staticmethod
+    def _get_resource_link_id(course_key: CourseKey, request: HttpRequest) -> str:
+        site = get_current_site(request)
+        return '{}-{}'.format(
+            site.domain,
+            str(course_key.make_usage_key('course', course_key.run)),
+        )
+
+    @staticmethod
+    def _get_result_sourcedid(context_id: str, resource_link_id: str, user_id: str) -> str:
+        return "{context}:{resource_link}:{user_id}".format(
+            context=context_id,
+            resource_link=resource_link_id,
+            user_id=user_id,
+        )
+
+    @staticmethod
+    def _get_context_title(course: CourseDescriptor) -> str:
+        return "{} - {}".format(
+            course.display_name_with_default,
+            course.display_org_with_default,
+        )
+
+    def _get_lti_config(self, course: CourseDescriptor) -> LtiConfiguration:
+        raise NotImplementedError
```

Comment:
Do we anticipate getting LTI configuration from other places beside the LtiConfiguration model?  Wondering if it just makes sense to implement this in the mixin, rather than delegating to the view to implement.  (I don't feel terribly strongly either way)

Reply:
I left it like this since for the LTI tab we need to get the config from the LtiConfiguration model stored in the tab, while for discussiosn it's linked via the discussion config model.